### PR TITLE
SageMaker Embedding Driver

### LINKED ISSUE #371: Add `AmazonSageMakerEmbeddingDriver`
Add `AmazonSageMakerEmbeddingDriver` to support SageMaker endpoints for generating embeddings.

### PR DESCRIPTION
**Pull request by collindutter**
Closes #371, #387

### CODE REVIEW

File: griptape/drivers/embedding/sagemaker_embedding_driver.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+import json
+from typing import Any
+
+import boto3
```
Comment (andrewfrench) [inline griptape/drivers/embedding/sagemaker_embedding_driver.py:6]: Elsewhere in the framework, like in bedrock_titan_embedding_driver.py, we conditionally `import boto3` under the `if TYPE_CHECKING` block like BaseEmbeddingModelDriver below, but it's not consistent. Should we do that here as well?

**Review by andrewfrench — COMMENTED** [review]
One small comment

File: griptape/drivers/embedding/sagemaker_embedding_driver.py (hunk 1 above)
Comment (collindutter) [reply griptape/drivers/embedding/sagemaker_embedding_driver.py]: Good catch, that would've broken things. Also updated the non-typed import of boto3 to use `import_optional_dependency`.

**Review by andrewfrench — APPROVED** [review]

**Review by andrewfrench — APPROVED** [review]

File: griptape/drivers/embedding/base_embedding_driver.py [hunk 2]
```diff
@@ -12,23 +12,25 @@
...
     """
     Attributes:
+        model: The name of the model to use.
         tokenizer: An instance of `BaseTokenizer` to use when calculating tokens.
-        chunker: An instance of `BaseChunker` to use when chunking long strings.
     """
 
-    tokenizer: BaseTokenizer = field(kw_only=True)
+    model: str = field(kw_only=True)
+    tokenizer: BaseTokenizer | None = field(kw_only=True)
```
Comment (vasinov) [inline griptape/drivers/embedding/base_embedding_driver.py:20]: 1. `Optional[BaseTokenizer]`
2. Should it have `None` as the default?

File: griptape/drivers/embedding/base_embedding_driver.py [hunk 3]
```diff
@@ -12,23 +12,25 @@
...
         return self.embed_string(artifact.to_text())
 
     def embed_string(self, string: str) -> list[float]:
         for attempt in self.retrying():
             with attempt:
-                if self.tokenizer.count_tokens(string) > self.tokenizer.max_tokens:
+                if self.tokenizer and self.tokenizer.count_tokens(string) > self.tokenizer.max_tokens:
                     return self._embed_long_string(string)
                 else:
                     return self.try_embed_chunk(string)
```
Comment (vasinov) [inline griptape/drivers/embedding/base_embedding_driver.py:36]: Is this meaningful behavior if the tokenizer is not specified? What use case does it enable?

**Review by vasinov — CHANGES_REQUESTED** [review]

File: griptape/drivers/embedding/base_embedding_driver.py (hunk 3 above)
Comment (collindutter) [reply griptape/drivers/embedding/base_embedding_driver.py:36]: A Tokenizer is only needed when embedding long strings that need to be chunked up. Some embedding models do not provide tokenization information (max tokens, tokenization algorithm, etc) so there is no great way to build a tokenizer. By making it optional, we hand the reigns to the developer to embed strings of a length they determine.

File: griptape/drivers/embedding/base_embedding_driver.py (hunk 2 above)
Comment (collindutter) [reply griptape/drivers/embedding/base_embedding_driver.py:20]: 1. This change comes from pyupgrade: https://github.com/asottile/pyupgrade?tab=readme-ov-file#pep-604-typing-rewrites
2. Yep.

**Review by vasinov — APPROVED** [review]

**Pull request merged**